### PR TITLE
Apm 228 auth and user

### PR DESCRIPTION
**Pull request by alexhawdon**
## Summary
* :sparkles: OAuth API Proxy
* :sparkles: Endpoints in PDS API Proxy secured

See the ticket and the design docs for a fuller understanding of these changes

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [x] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [x] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [x] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<AssignMessage async="false" continueOnError="false" enabled="true" name="AssignMessage.SetCorsHeaders">
+  <Set>
+    <Headers>
+      <Header name="Access-Control-Allow-Origin">{request.header.origin}</Header>
+      <Header name="Access-Control-Allow-Headers">origin, x-requested-with, accept, content-type, authorization</Header>
+      <Header name="Access-Control-Max-Age">3628800</Header>
```
Comment (Sourceless) [inline identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml]: This is an extremely long expiry time - let's keep it fairly short. Absolute max 1 day, 5 mins-2 hours is fairly standard in most browsers.

File: identity-v1/apiproxy/resources/jsc/create-cache-entry.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+var1 = context.getVariable('request.queryparam.client_id');
```
Comment (Sourceless) [inline identity-v1/apiproxy/resources/jsc/create-cache-entry.js]: Rename variable to what they are

File: specification/patient-demographics-service-api.yaml [hunk 3]
```diff
@@ -727,6 +731,30 @@ components:
...
+      name: Authorization
+      description: |
+        An OAuthV2 access token presented in Bearer format.
+
+        Note: This parameter is required unless interacting with the Sandbox.
+      required: false
+      schema:
+        type: string
+        format: '^Bearer\:\ [[:ascii:]]+$'
+        example: 'Bearer: g1112R_ccQ1Ebbb4gtHBP1aaaNM'
```
Comment (Sourceless) [inline specification/patient-demographics-service-api.yaml]: Is this an apigee thing? There shouldn't be a `:` after `Bearer`

https://tools.ietf.org/html/rfc6750

File: identity-v1/apiproxy/policies/AssignMessage.GenerateState.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<AssignMessage async="false" continueOnError="false" enabled="true" name="AssignMessage.GenerateState">
+    <AssignVariable>
+        <Name>apigee.state</Name>
+        <Template>{randomLong(args)}</Template>
```
[the remark this answers is not in the document]
Comment (alexhawdon) [reply identity-v1/apiproxy/policies/AssignMessage.GenerateState.xml]: This is normal, it's how a 'state' variable is generated and cached for the coke-for-token exchange between Apigee and the Identity Provider

File: identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml (hunk 1 above)
Comment (alexhawdon) [reply identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml]: Per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age I'll take this down to 10 minutes (Chromium's old maximum)

File: specification/patient-demographics-service-api.yaml (hunk 3 above)
Comment (alexhawdon) [reply specification/patient-demographics-service-api.yaml]: Nope that's just plain wrong on my part! Good catch

File: identity-v1/apiproxy/policies/AssignMessage.GenerateState.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sourceless) [reply identity-v1/apiproxy/policies/AssignMessage.GenerateState.xml]: ah right now i've looked properly, randomLong is 160 bit cryptographically secure. That's (almost certainly) likely to never collide, so is fine

File: identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml (hunk 1 above)
Comment (Sourceless) [reply identity-v1/apiproxy/policies/AssignMessage.SetCorsHeaders.xml]: :+1:

**Pull request merged**